### DISCUSSION TITLE
API Designer

### DISCUSSION THREAD

**Discussion by VaZark** [directus/directus#12053] [Feature Requests]
From my experience so far, Directus is very much a DB-first service. The DB defines the APIs and their structure. 

What do you think about having an API designer as an intermediate layer? The purpose of this layer is similar to a serializer which filters and/or collects data from one or more tables and exposes them as a REST API (I do not have much idea about GraphQL). This can be made more generic and essential to core logic by subclassing the auto-generated API for tables to be instances of this layer.

With these 2 layers of controls, it allows backend devs to create custom endpoints only for complex business logic that transforms the data in real-time.

**Comment by benhaynes** [comment]
We have something big in the works that will give us a lot of this functionality. Give us a few weeks to get that out, and then we can discuss this further. Or, ping us on Discord and we might be able to discuss the new feature privately.

**Reply by VaZark** [reply]
That sounds fantastic. I'll ping this discussion after the next major release then.

**Comment by VaZark** [comment]
Let me just start by saying that the Data Flow interface looks absolutely amazing. Moreover, the Dataflow having Webhook support out of the box is fantastic.

Going through the PR (couldn't build it :sweat_smile:) gave me a couple of ideas. I'll bounce them here and maybe some will interest the team too.

1. Option to add the JS code directly from the app for admins with read-only permission for select users
3. Allow extensions to register to DataFlows and/or register custom code that can be used with data flows
2. Option to reimplement the Custom API as a data flow too. (Does webhook have support for fixed schema??)

_API Designer_
1. Re-use the data flow interface to show APIs that are automatically generated or make them a subclass of operations
2. Allow option to select schema of the API from admin (show/hide fields, expand foreign key to show parent fields, etc.)

**Reply by benhaynes** [reply]
CC @rijkvanzanten @nickrum

**Reply by benhaynes** [reply]
Thanks!! One thing that I'd like to get into Data Flows is an "alias" feature, so that you can give each flow a dedicated URL that is more semantic than a UUID. I think this could be pretty powerful as a no-code builder for API endpoints too. :thum

**Reply by VaZark** [reply]
I second that. Having a semantic URL for webhooks + `operations` for custom code will also expose the logic on the Dataflow interface and make Custom API redundant for most use-cases. 

With that, Directus will have all of its custom endpoints visible on the no-code app which makes it easier to monitor the APIs when dealing with adding new extensions.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
